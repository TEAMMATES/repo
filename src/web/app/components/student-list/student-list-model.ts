import { ColumnData, SortableTableCellData } from '../sortable-table/sortable-table.component';

export enum StudentListColumns {
  SECTION,
  TEAM,
  STUDENT_NAME,
  STATUS,
  EMAIL,
  ACTIONS,
}

export const StudentListColumnNames = new Map<StudentListColumns, string>([
  [StudentListColumns.SECTION, 'Section'],
  [StudentListColumns.TEAM, 'Team'],
  [StudentListColumns.STUDENT_NAME, 'Student Name'],
  [StudentListColumns.STATUS, 'Status'],
  [StudentListColumns.EMAIL, 'Email'],
  [StudentListColumns.ACTIONS, 'Action(s)'],
]);

export interface StudentListColumnData extends ColumnData {
  columnType?: StudentListColumns;
}

export interface StudentListRowData extends SortableTableCellData {
  columnType?: StudentListColumns;
  section?: string;
  team?: string;
  studentName?: string;
  status?: string;
  email?: string;
  actions?: string;
}
