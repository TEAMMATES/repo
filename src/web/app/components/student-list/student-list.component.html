<div class="table-responsive">
  <tm-sortable-table
    [columns]="columnData"
    [rows]="rowData"
    
    [initialSortBy]="tableSortBy"
    
    [sortOrder]="tableSortOrder"
    [headerColorScheme]="headerColorScheme"
    [customHeaderStyle]="customHeaderStyle"
    (sortEvent)="sortStudentListEventHandler($event)">
  </tm-sortable-table>
</div>
