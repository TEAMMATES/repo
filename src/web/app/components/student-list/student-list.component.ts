import { Component, EventEmitter, Input, Output, OnInit } from '@angular/core';
import { NgbModalRef } from '@ng-bootstrap/ng-bootstrap';
import { CourseService } from '../../../services/course.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { JoinState, MessageOutput, Student } from '../../../types/api-output';
import { SortBy, SortOrder } from '../../../types/sort-properties';
import { ErrorMessageOutput } from '../../error-message-output';
import { SimpleModalType } from '../simple-modal/simple-modal-type';
import {
    ColumnData,
    SortableEvent,
    SortableTableCellData,
    SortableTableHeaderColorScheme,
} from '../sortable-table/sortable-table.component';
import { StudentListActionsComponent } from './student-list-action-cell.component';
import {
    StudentListColumnData,
    StudentListRowData,
    StudentListColumns,
} from './student-list-model';

/**
 * Model of row of student data containing details about a student and their section.
 */
export interface StudentListRowModel {
  student: Student;
  isAllowedToViewStudentInSection: boolean;
  isAllowedToModifyStudent: boolean;
}

/**
 * A table displaying a list of students from a course, with buttons to view/edit/delete students etc.
 */
@Component({
  selector: 'tm-student-list',
  templateUrl: './student-list.component.html',
  styleUrls: ['./student-list.component.scss'],
})
export class StudentListComponent implements OnInit {
  @Input() courseId: string = '';
  @Input() useGrayHeading: boolean = true;
  @Input() listOfStudentsToHide: string[] = [];
  @Input() isHideTableHead: boolean = false;
  @Input() enableRemindButton: boolean = false;
  @Input() isActionButtonsEnabled: boolean = true;
  @Input() students: StudentListRowModel[] = [];
  @Input() tableSortBy: SortBy = SortBy.NONE;
  @Input() tableSortOrder: SortOrder = SortOrder.ASC;
  @Input() searchString: string = '';
  @Input() headerColorScheme: SortableTableHeaderColorScheme = SortableTableHeaderColorScheme.OTHERS;
  @Input() customHeaderStyle: string = 'bg-light';

  @Output() removeStudentFromCourseEvent: EventEmitter<string> = new EventEmitter();
  @Output() sortStudentListEvent: EventEmitter<SortableEvent> = new EventEmitter();

  rowData: SortableTableCellData[][] = [];
  columnData: ColumnData[] = [];

  // enum
  SortBy: typeof SortBy = SortBy;
  SortOrder: typeof SortOrder = SortOrder;
  JoinState: typeof JoinState = JoinState;
  StudentListColumns: typeof StudentListColumns = StudentListColumns;
  SortableTableHeaderColorScheme: typeof SortableTableHeaderColorScheme = SortableTableHeaderColorScheme;

  constructor(private statusMessageService: StatusMessageService,
              private courseService: CourseService,
              private simpleModalService: SimpleModalService) {
  }

  /**
   * Returns whether this course are divided into sections
   */
  hasSection(): boolean {
    return (this.students.some((studentModel: StudentListRowModel) =>
        studentModel.student.sectionName !== 'None'));
  }

  ngOnInit(): void {
    this.setColumnData();
    this.setRowData();
    this.setHeaderStyle();
  }

  setHeaderStyle(): void {
    this.customHeaderStyle = this.useGrayHeading ? 'bg-light' : 'bg-info';
  }

  createColumnData(config: StudentListColumnData): ColumnData[] {

    const columnData: ColumnData = {
      header: config.header,
      ...(config.sortBy && { sortBy: config.sortBy }),
      ...(config.headerToolTip && { headerToolTip: config.headerToolTip }),
      ...(config.alignment && { alignment: config.alignment }),
      ...(config.headerClass && { headerClass: config.headerClass }),
    };

    return [columnData];
  }

  setColumnData(): void {
    this.columnData = [
        ...this.createColumnData({
            columnType: StudentListColumns.SECTION,
            header: 'Section',
            sortBy: SortBy.SECTION_NAME,
        }),
        ...this.createColumnData({
            columnType: StudentListColumns.TEAM,
            header: 'Team',
            sortBy: SortBy.TEAM_NAME,
        }),
        ...this.createColumnData({
            columnType: StudentListColumns.STUDENT_NAME,
            header: 'Student Name',
            sortBy: SortBy.RESPONDENT_NAME,
        }),
        ...this.createColumnData({
            columnType: StudentListColumns.STATUS,
            header: 'Status',
            sortBy: SortBy.JOIN_STATUS,
        }),
        ...this.createColumnData({
            columnType: StudentListColumns.EMAIL,
            header: 'Email',
            sortBy: SortBy.RESPONDENT_EMAIL,
        }),
        ...this.createColumnData({ columnType: StudentListColumns.ACTIONS, header: 'Action(s)', alignment: 'center' }),
    ];
  }

  createRowData(config: StudentListRowData): SortableTableCellData[] {
  
    const rowData: SortableTableCellData = {
      ...(config.value && { value: config.value }),
      ...(config.displayValue && { displayValue: config.displayValue }),
      ...(config.customComponent && { customComponent: config.customComponent }),
      ...(config.style && { style: config.style }),
    };

    return [rowData];
  }

  setRowData(): void {
    this.rowData = this.students.map((studentModel: StudentListRowModel) => {
      const rowData: SortableTableCellData[] = [
        ...this.createRowData({
          columnType: StudentListColumns.SECTION,
          value: studentModel.student.sectionName,
        }),
        ...this.createRowData({
          columnType: StudentListColumns.TEAM,
          value: studentModel.student.teamName,
        }),
        ...this.createRowData({
          columnType: StudentListColumns.STUDENT_NAME,
          value: studentModel.student.name,
        }),
        ...this.createRowData({
          columnType: StudentListColumns.STATUS,
          value: studentModel.student.joinState === JoinState.JOINED ? 'Joined' : 'Yet to Join',
        }),
        ...this.createRowData({
          columnType: StudentListColumns.EMAIL,
          value: studentModel.student.email,
        }),
        ...this.createRowData(this.createActionsCell(studentModel)),
      ];

      return rowData;
    });
  }

  createActionsCell(studentModel: StudentListRowModel): SortableTableCellData {
    const actionsCell: StudentListRowData = {
      columnType: StudentListColumns.ACTIONS,
      customComponent: {
      component: StudentListActionsComponent,
      componentData: (idx: number) => ({
        idx,
        courseId: this.courseId,
        email: studentModel.student.email,
        hasJoined: studentModel.student.joinState === JoinState.JOINED,
        removeStudentFromCourse: () => this.openDeleteModal(studentModel),
        remindStudentFromCourse: () => this.openRemindModal(studentModel),
        instructorPrivileges: {
            canModifyStudent: studentModel.isAllowedToModifyStudent,
            canViewStudentInSections: studentModel.isAllowedToViewStudentInSection,
        },
        enableRemindButton: studentModel.student.joinState === JoinState.NOT_JOINED,
        isActionButtonsEnabled: this.isActionButtonsEnabled,
      }),
      },
    };

    return actionsCell;
  }

  /**
   * Function to be passed to ngFor, so that students in the list is tracked by email
   */
  trackByFn(_index: number, item: StudentListRowModel): any {
    return item.student.email;
  }

  /**
   * Open the student email reminder modal.
   */
  openRemindModal(studentModel: StudentListRowModel): void {
    const modalContent: string = `Usually, there is no need to use this feature because
          TEAMMATES sends an automatic invite to students at the opening time of each session.
          Send a join request to <strong>${studentModel.student.email}</strong> anyway?`;
    const modalRef: NgbModalRef = this.simpleModalService.openConfirmationModal(
        'Send join request?', SimpleModalType.INFO, modalContent);
    modalRef.result.then(() => {
      this.remindStudentFromCourse(studentModel.student.email);
    }, () => {});
  }

  /**
   * Open the delete student confirmation modal.
   */
  openDeleteModal(studentModel: StudentListRowModel): void {
    const modalContent: string = `Are you sure you want to remove <strong>${studentModel.student.name}</strong> `
        + `from the course <strong>${this.courseId}?</strong>`;
    const modalRef: NgbModalRef = this.simpleModalService.openConfirmationModal(
        `Delete student <strong>${studentModel.student.name}</strong>?`, SimpleModalType.DANGER, modalContent);
    modalRef.result.then(() => {
      this.removeStudentFromCourse(studentModel.student.email);
    }, () => {});
  }

  /**
   * Remind the student from course.
   */
  remindStudentFromCourse(studentEmail: string): void {
    this.courseService.remindStudentForJoin(this.courseId, studentEmail)
      .subscribe({
        next: (resp: MessageOutput) => {
          this.statusMessageService.showSuccessToast(resp.message);
        },
        error: (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      });
  }

  /**
   * Removes the student from course.
   */
  removeStudentFromCourse(studentEmail: string): void {
    this.removeStudentFromCourseEvent.emit(studentEmail);
  }

  /**
   * Determines which row in the studentTable should be hidden.
   */
  isStudentToHide(studentEmail: string): boolean {
    return this.listOfStudentsToHide.indexOf(studentEmail) > -1;
  }

  /**
   * Sorts the student list
   */
  sortStudentListEventHandler(event: { sortBy: SortBy, sortOrder: SortOrder }): void {
    this.sortStudentListEvent.emit(event);
  }

  getAriaSort(by: SortBy): String {
    if (by !== this.tableSortBy) {
      return 'none';
    }
    return this.tableSortOrder === SortOrder.ASC ? 'ascending' : 'descending';
  }
}
