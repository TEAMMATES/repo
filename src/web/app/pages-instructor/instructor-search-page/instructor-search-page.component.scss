.bg-light-blue {
    background-color: #007bff33 !important;
}

.border-light-blue {
    border-color: #007bff33 !important;
}

.text-dark-blue {
    color: var(--blue) !important;
}
