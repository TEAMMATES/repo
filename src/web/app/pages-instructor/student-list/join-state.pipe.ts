import { Pipe, PipeTransform } from '@angular/core';
import { JoinState } from "../../../types/api-output";

@Pipe({
  name: 'joinState'
})
export class JoinStatePipe implements PipeTransform {

  transform(joinState: JoinState): any {
    switch(joinState) {
      case (JoinState.JOINED):
        return 'Joined';
      case (JoinState.NOT_JOINED):
        return 'Yet to Join';
      default:
        return 'Unknown';
    }
  }

}
