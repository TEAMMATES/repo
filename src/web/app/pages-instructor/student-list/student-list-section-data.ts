import { JoinState } from '../../../types/api-output';

/**
 * Contains a list of students in a section.
 */
export interface StudentListSectionData {
  sectionName: string;
  isAllowedToViewStudentInSection: boolean;
  isAllowedToModifyStudent: boolean;
  students: StudentListStudentData[];
}

/**
 * Contains details about a student to be displayed in the list.
 */
export interface StudentListStudentData {
  name: string;
  email: string;
  status: JoinState;
  team: string;
  photoUrl?: string;
}

/**
 * Sort criteria for the students table.
 */
export enum SortBy {
  /**
   * Nothing.
   */
  NONE,

  /**
   * Section Name.
   */
  SECTION_NAME,

  /**
   * Team name.
   */
  TEAM_NAME,

  /**
   * Student Name.
   */
  STUDENT_NAME,

  /**
   * Status.
   */
  STATUS,

  /**
   * Email.
   */
  EMAIL,
}

/**
 * Sort order for the students table.
 */
export enum SortOrder {
  /**
   * Descending sort order.
   */
  DESC,

  /**
   * Ascending sort order
   */
  ASC,
}
