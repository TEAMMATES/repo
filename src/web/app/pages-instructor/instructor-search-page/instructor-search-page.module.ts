import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { StudentListModule } from '../student-list/student-list.module';
import { InstructorSearchPageComponent } from './instructor-search-page.component';

/**
 * Module for instructor search page.
 */
@NgModule({
  declarations: [
    InstructorSearchPageComponent,
  ],
  exports: [
    InstructorSearchPageComponent,
  ],
  imports: [
    CommonModule,
    StudentListModule,
    FormsModule,
    RouterModule,
  ],
})
export class InstructorSearchPageModule { }
