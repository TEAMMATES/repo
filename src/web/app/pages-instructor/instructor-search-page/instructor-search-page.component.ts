import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { finalize } from 'rxjs/operators';
import { LoadingBarService } from '../../../services/loading-bar.service';
import {
  InstructorSearchResult,
  SearchService,
} from '../../../services/search.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { ErrorMessageOutput } from '../../error-message-output';
import { StudentListSectionData } from '../student-list/student-list-section-data';

/**
 * Data object for communication with the child student result component
 */
export interface SearchStudentsTable {
  courseId: string;
  sections: StudentListSectionData[];
}

/**
 * Instructor search page.
 */
@Component({
  selector: 'tm-instructor-search-page',
  templateUrl: './instructor-search-page.component.html',
  styleUrls: ['./instructor-search-page.component.scss'],
})
export class InstructorSearchPageComponent implements OnInit {
  searchKey: string = '';
  studentTables: SearchStudentsTable[] = [];

  constructor(
    private route: ActivatedRoute,
    private statusMessageService: StatusMessageService,
    private loadingBarService: LoadingBarService,
    private searchService: SearchService,
  ) {}

  ngOnInit(): void {
    this.route.queryParams.subscribe((queryParams: any) => {
      if (queryParams.studentSearchkey) {
        this.searchKey = queryParams.studentSearchkey;
      }
      if (this.searchKey) {
        this.search(this.searchKey);
      }
    });
  }

  /**
   * Searches for students and questions/responses/comments matching the search query.
   */
  search(searchKey: string): void {
    this.loadingBarService.showLoadingBar();
    this.searchService
      .searchInstructor(searchKey)
      .pipe(finalize(() => this.loadingBarService.hideLoadingBar()))
      .subscribe(
        (resp: InstructorSearchResult) => {
          this.studentTables = resp.searchStudentsTables;
          const hasStudents: boolean = !!(
            this.studentTables && this.studentTables.length
          );
          if (!hasStudents) {
            this.statusMessageService.showWarningMessage('No results found.');
          }
        },
        (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorMessage(resp.error.message);
        },
      );
  }
}
