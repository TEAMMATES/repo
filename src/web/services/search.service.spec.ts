import { TestBed } from '@angular/core/testing';

import { HttpClientTestingModule } from '@angular/common/http/testing';
import { of } from 'rxjs';
import { SearchStudentsTable } from '../app/pages-instructor/instructor-search-page/instructor-search-page.component';
import { StudentListSectionData } from '../app/pages-instructor/student-list/student-list-section-data';
import { ResourceEndpoints } from '../types/api-endpoints';
import { InstructorPrivilege, JoinState, Student, Students } from '../types/api-output';
import { HttpRequestService } from './http-request.service';
import { SearchService } from './search.service';

describe('SearchService', () => {
  let spyHttpRequestService: any;
  let service: SearchService;

  const mockStudents: Students = {
    students: [
      {
        email: 'alice@example.com',
        courseId: 'CS3281',
        name: 'Alice',
        joinState: JoinState.JOINED,
        teamName: 'Team 1',
        sectionName: 'Section 1',
      },
      {
        email: 'bob@example.com',
        courseId: 'CS3281',
        name: 'Bob',
        joinState: JoinState.JOINED,
        teamName: 'Team 1',
        sectionName: 'Section 1',
      },
      {
        email: 'chloe@example.com',
        courseId: 'CS3281',
        name: 'Chloe',
        joinState: JoinState.JOINED,
        teamName: 'Team 1',
        sectionName: 'Section 2',
      },
      {
        email: 'david@example.com',
        courseId: 'CS3282',
        name: 'David',
        joinState: JoinState.JOINED,
        teamName: 'Team 1',
        sectionName: 'Section 2',
      },
    ],
  };
  let coursesWithSections: SearchStudentsTable[];

  beforeEach(() => {
    spyHttpRequestService = {
      get: jest.fn(),
      post: jest.fn(),
      put: jest.fn(),
      delete: jest.fn(),
    };
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule],
      providers: [
        { provide: HttpRequestService, useValue: spyHttpRequestService },
      ],
    });
    service = TestBed.get(SearchService);
    coursesWithSections = service.getCoursesWithSections(mockStudents);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  it('should execute GET when searching for students', () => {
    service.getStudents('Alice');
    const paramMap: { [key: string]: string } = {
      searchkey: 'Alice',
    };
    expect(spyHttpRequestService.get).toHaveBeenCalledWith(
      '/search/students',
      paramMap,
    );
  });

  it('should parse students into courses with sections correctly', () => {
    const { students }: { students: Student[] } = mockStudents;

    // Number of courses should match
    expect(coursesWithSections.length).toEqual(
      Array.from(new Set(students.map((s: Student) => s.courseId))).length,
    );

    // Number of sections in a course should match
    expect(
      coursesWithSections.filter((t: SearchStudentsTable) => t.courseId === students[0].courseId)[0]
        .sections.length,
    ).toEqual(
      Array.from(
        new Set(
          students
            .filter((s: Student) => s.courseId === students[0].courseId)
            .map((s: Student) => s.sectionName),
        ),
      ).length,
    );

    // Number of students in a section should match
    expect(
      coursesWithSections
        .filter((t: SearchStudentsTable) => t.courseId === students[0].courseId)[0]
        .sections.filter((s: StudentListSectionData) => s.sectionName === students[0].sectionName)[0]
        .students.length,
    ).toEqual(
      students.filter((s: Student) => s.sectionName === students[0].sectionName).length,
    );
  });

  it('should execute GET when fetching privileges', () => {
    spyHttpRequestService.get.mockImplementation((endpoint: string) => {
      expect(endpoint).toEqual(ResourceEndpoints.INSTRUCTOR_PRIVILEGE);
      return of<InstructorPrivilege>({
        canModifyCourse: true,
        canModifySession: true,
        canModifyStudent: true,
        canModifyInstructor: true,
        canViewStudentInSections: true,
        canModifySessionCommentsInSections: true,
        canViewSessionInSections: true,
        canSubmitSessionInSections: true,
      });
    });
    service.getPrivileges(coursesWithSections);

    for (const course of coursesWithSections) {
      for (const section of course.sections) {
        expect(spyHttpRequestService.get).toHaveBeenCalledWith(
          ResourceEndpoints.INSTRUCTOR_PRIVILEGE,
          {
            courseid: course.courseId,
            sectionname: section.sectionName,
          },
        );
      }
    }
  });

  it('should combine privileges and course data correctly', () => {
    const basePrivilege: InstructorPrivilege = {
      canModifyCourse: true,
      canModifySession: true,
      canModifyStudent: true,
      canModifyInstructor: true,
      canViewStudentInSections: true,
      canModifySessionCommentsInSections: true,
      canViewSessionInSections: true,
      canSubmitSessionInSections: true,
    };
    const mockPrivileges: InstructorPrivilege[] = [
      basePrivilege,
      {
        ...basePrivilege,
        canViewStudentInSections: false,
        canModifyStudent: true,
      },
      {
        ...basePrivilege,
        canViewStudentInSections: true,
        canModifyStudent: false,
      },
    ];
    service.combinePrivileges([coursesWithSections, mockPrivileges]);

    const course1Section1: StudentListSectionData = coursesWithSections[0].sections[0];
    expect(course1Section1.isAllowedToViewStudentInSection).toEqual(true);
    expect(course1Section1.isAllowedToModifyStudent).toEqual(true);

    const course1Section2: StudentListSectionData = coursesWithSections[0].sections[1];
    expect(course1Section2.isAllowedToViewStudentInSection).toEqual(false);
    expect(course1Section2.isAllowedToModifyStudent).toEqual(true);

    const course2Section1: StudentListSectionData = coursesWithSections[1].sections[0];
    expect(course2Section1.isAllowedToViewStudentInSection).toEqual(true);
    expect(course2Section1.isAllowedToModifyStudent).toEqual(false);

    expect(mockPrivileges.length).toEqual(0);
  });
});
