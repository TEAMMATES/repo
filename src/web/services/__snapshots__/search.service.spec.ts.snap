// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchService should join instructors accurately when calling as admin 1`] = `
Object {
  "courseId": "dog.gma-demo",
  "courseJoinLink": "http://localhost:4200/web/join?entitytype=instructor",
  "courseName": "Sample Course 101",
  "email": "dog@gmail.com",
  "googleId": "test@example.com",
  "homePageLink": "/web/instructor/home?user=test%40example.com",
  "institute": "",
  "manageAccountLink": "/web/admin/accounts?instructorid=test%40example.com",
  "name": "Hi",
  "showLinks": false,
}
`;
